xarray/core/utils.py

```diff
@@ -247,9 +249,19 @@ def is_list_like(value: Any) -> bool:
     return isinstance(value, list) or isinstance(value, tuple)
 
 
-def is_array_like(value: Any) -> bool:
+def is_duck_array(value: Any) -> bool:
     return (
-        hasattr(value, "ndim") and hasattr(value, "shape") and hasattr(value, "dtype")
+        hasattr(value, "ndim")
+        and hasattr(value, "shape")
+        and hasattr(value, "dtype")
+        and hasattr(value, "__array_function__")
+        and hasattr(value, "__array_ufunc__")
+    )
+
+
+def is_duck_array_or_ndarray(value: Any) -> bool:
+    return is_duck_array(value) or (
+        not IS_NEP18_ACTIVE and isinstance(value, np.ndarray)
     )
```

Comment:
Is there any circumstance under which we would *not* want to consider NumPy arrays duck arrays? Even if they don't define `__array_function__`, it is always fine to use functions from the `numpy` namespace on them directly. It doesn't matter if NEP-18 is active or not.

Given that we expect _most_ arguments to this function to be NumPy arrays, there's another reason include a check for NumPy arrays explicitly as well: it's much faster, e.g., this version checks a NumPy array in 170 ns vs 550 ns:
```python
def is_duck_array(value):
    if isinstance(value, np.ndarray):
        # much faster than checking for attributes
        # also, NumPy prior to 1.17 does not have an __array_function__ method
        return True
    return (
        hasattr(value, "ndim")
        and hasattr(value, "shape")
        and hasattr(value, "dtype")
        and hasattr(value, "__array_function__")
        and hasattr(value, "__array_ufunc__")
    )
```

Reply:
I can't think of a situation where NumPy arrays shouldn't be considered duck arrays. Your solution is also more simple than having another function just for these cases. I'll go ahead and make this change.